crates/bevy_pbr/src/wireframe.rs

```diff
@@ -41,158 +31,101 @@ impl Plugin for WireframePlugin {
+#[derive(Component, Debug, Clone, Default, Reflect)]
 #[reflect(Component, Default)]
 pub struct Wireframe;
 
+/// Configuration resource for [`WireframePlugin`].
 #[derive(Resource, Debug, Clone, Default, ExtractResource, Reflect)]
 #[reflect(Resource)]
 pub struct WireframeConfig {
-    /// Whether to show wireframes for all meshes. If `false`, only meshes with a [Wireframe] component will be rendered.
+    /// Whether to show wireframes for all meshes.
+    /// If `false`, only meshes with a [`Wireframe`] component will be rendered.
     pub global: bool,
 }
 
-#[derive(Resource, Clone)]
-pub struct WireframePipeline {
-    mesh_pipeline: MeshPipeline,
-    shader: Handle<Shader>,
-}
-impl FromWorld for WireframePipeline {
-    fn from_world(render_world: &mut World) -> Self {
-        WireframePipeline {
-            mesh_pipeline: render_world.resource::<MeshPipeline>().clone(),
-            shader: WIREFRAME_SHADER_HANDLE.typed(),
-        }
+/// Applies the wireframe material to any mesh with a [`Wireframe`] component.
+fn apply_material(
+    mut commands: Commands,
+    mut materials: ResMut<Assets<WireframeMaterial>>,
+    wireframes: Query<Entity, (With<Wireframe>, Without<Handle<WireframeMaterial>>)>,
+) {
+    for e in &wireframes {
+        commands
+            .entity(e)
+            .insert(materials.add(WireframeMaterial {}));
     }
 }
 
-impl SpecializedMeshPipeline for WireframePipeline {
-    type Key = MeshPipelineKey;
+/// Applies or removes a wireframe material on any mesh without a [`Wireframe`] component.
+#[allow(clippy::type_complexity)]
+fn apply_global(
+    mut commands: Commands,
+    config: Res<WireframeConfig>,
+    mut materials: ResMut<Assets<WireframeMaterial>>,
+    meshes_without_material: Query<
+        Entity,
+        (
+            With<Handle<Mesh>>,
+            Without<Wireframe>,
+            Without<Handle<WireframeMaterial>>,
+        ),
+    >,
+    meshes_with_material: Query<
+        Entity,
+        (
+            With<Handle<Mesh>>,
+            Without<Wireframe>,
+            With<Handle<WireframeMaterial>>,
+        ),
+    >,
+) {
+    if !config.is_changed() {
+        return;
+    }
 
-    fn specialize(
-        &self,
-        key: Self::Key,
-        layout: &MeshVertexBufferLayout,
-    ) -> Result<RenderPipelineDescriptor, SpecializedMeshPipelineError> {
-        let mut descriptor = self.mesh_pipeline.specialize(key, layout)?;
-        descriptor.vertex.shader = self.shader.clone_weak();
-        descriptor
-            .vertex
-            .shader_defs
-            .push("MESH_BINDGROUP_1".into());
-        descriptor.fragment.as_mut().unwrap().shader = self.shader.clone_weak();
-        descriptor.primitive.polygon_mode = PolygonMode::Line;
-        descriptor.depth_stencil.as_mut().unwrap().bias.slope_scale = 1.0;
-        Ok(descriptor)
+    if config.global {
+        let global_material = materials.add(WireframeMaterial {});
```

Comment:
Could this handled be cached instead to enable more batching?

Reply:
This may be an issue once I add color support but for this PR I decided to use a shared handle here.